### PR TITLE
Make git pulls less error prone due to history changes.

### PR DESCRIPTION
**Pull request by josepht**
Bugs/1614520

### CODE REVIEW

File: snapcraft/internal/sources/_git.py [hunk 1]
```diff
@@ -49,6 +49,11 @@ def _pull_existing(self):
...
             refspec = self.source_commit
 
+        subprocess.check_call([self.command, '-C', self.source_dir,
+                               'fetch', '--recurse-submodules=yes'])
+        subprocess.check_call([self.command, '-C', self.source_dir,
+                               'reset', '--hard', refspec])
+
         # Pull changes to this repository and any submodules.
         subprocess.check_call([self.command, '-C', self.source_dir,
                                'pull', '--recurse-submodules=yes',
```
Comment (kyrofa) [inline snapcraft/internal/sources/_git.py]: Wait wait, why are you pulling after the fetch/reset?

Comment (josepht) [inline snapcraft/internal/sources/_git.py]: Does the reset recurse over the submodules?  I think I'll need to do something along the lines of what is suggested here: http://stackoverflow.com/a/17871677

Comment (kyrofa) [inline snapcraft/internal/sources/_git.py]: Hmm, that's probably a good call, but some good testing is in order.

Comment (josepht) [inline snapcraft/internal/sources/_git.py]: Updated and test added.

**Comment by sergiusens** [comment]
leaving the call to you as you are going to do repull/rebuild

File: snapcraft/internal/sources/_git.py [hunk 2]
```diff
@@ -49,14 +53,18 @@ def _pull_existing(self):
...
         subprocess.check_call([self.command, '-C', self.source_dir,
-                               'pull', '--recurse-submodules=yes',
-                               self.source, refspec])
+                               'reset', '--hard', reset_spec], **self.kwargs)
 
         # Merge any updates for the submodules (if any).
         subprocess.check_call([self.command, '-C', self.source_dir,
-                              'submodule', 'update'])
+                              'submodule', 'update', '--remote'],
+                              **self.kwargs)
```
Comment (kyrofa) [inline snapcraft/internal/sources/_git.py]: Should this include `--recursive` to recursively make sure submodules within submodules are up-to-date?

Comment (josepht) [inline snapcraft/internal/sources/_git.py]: Updated

File: snapcraft/tests/sources/test_git.py [hunk 3]
```diff
@@ -161,3 +174,126 @@ def test_init_with_source_tag_and_commit_raises_exception(self):
...
+            shutil.rmtree(dir)
+
+    def clean_dir(self, dir):
+        self.rm_dir(dir)
+        os.mkdir(dir)
+        self.addCleanup(self.rm_dir, dir)
+
+    def clone_repo(self, repo, tree):
+        self.clean_dir(tree)
+        self.call(['git', 'clone', repo, tree])
```
Comment (sergiusens) [inline snapcraft/tests/sources/test_git.py]: this missed a `Build-Depends` on `git` in `debian/control` :cry:

**Pull request merged**